moveit_core/robot_state/src/robot_state.cpp

```diff
@@ -87,17 +87,25 @@ RobotState::~RobotState()
 
 void RobotState::allocMemory()
 {
+  static_assert((sizeof(Eigen::Isometry3d) / EIGEN_MAX_ALIGN_BYTES) * EIGEN_MAX_ALIGN_BYTES ==
```

Comment:
Is this really required since `EIGEN_MAX_ALIGN_BYTES` must be a power of 2?
Eigen even performs a check for that https://eigen.tuxfamily.org/dox/Constants_8h_source.html

Reply:
This assert is testing that `sizeof(Eigen::Isometry3d)` is a multiple of `EIGEN_MAX_ALIGN_BYTES`.
That's not ensured by `EIGEN_MAX_ALIGN_BYTES` being a power of 2.

Reply:
I thought that since Isometry3d has a fixed size of 128 bytes this always holds true because the highest allowed value of `EIGEN_MAX_ALIGN_BYTES` is 128. Or am I missing something?
The only case where the assertion doesn't hold true would be if `EIGEN_MAX_ALIGN_BYTES` is set to 0 which would break the condition.

Reply:
Exactly, and this condition just ensures that the size of Isometry3d doesn't change to e.g. 129 bytes, because a bool flag is added as member variable in future.

Reply:
I see! Eigen's aligned_allocator should handle it in that case but this is just nit-picking.

Reply:
I'm not sure `Eigen::aligned_allocator<T>` handles that case. Although I like the idea, to simply use the `Eigen::aligned_allocator` here, I didn't look into the details yet. 
One issue that immediately pops up, when trying to do so (and I actually tried yesterday), is which datatype should be passed to the allocator. As we use the allocated memory for several data items (of different type), that's not trivial. I think, one should use byte and specify the number of bytes required.
However, then I wasn't sure anymore how Eigen::aligned_allocator will actually handle alignment. Is it related to the `sizeof(T)`?

Reply:
The `Eigen::aligned_allocator` only supports Eigen types so we could only call it on the Isometry3d part. Internally it is pretty much the same implementation (should be `handmade_aligned_alloc()` in https://eigen.tuxfamily.org/dox/Memory_8h_source.html ) plus some sanity checks.
We would then have to add the variables part with `realloc()` so this would produce some overhead.

Reply:
I shortly looked into Eigen's memory management and stumbled upon this line in `Eigen::internal::handmade_aligned_malloc()`:
```c++
*(reinterpret_cast<void**>(aligned) - 1) = original;
```
They store the original pointer before the aligned one, which makes sense to properly cleanup the original pointer later on.
However, if the original pointer _is already aligned_, there is no space before to store the pointer.
This should actually corrupt memory in many cases, shouldn't it? Do I miss something?

Reply:
You are totally right. However, `handmade_aligned_malloc()` is called by `aligned_malloc()` which checks a parameter `EIGEN_MALLOC_ALREADY_ALIGNED` before that so that case should be prevented. There are actually a lot of preprocessor directives included that handle different system architectures.